src/Documents/Session/IDocumentQuery.ts

```diff
@@ -1,50 +1,91 @@
+import {QueryResultsCallback, EntityCallback, EntitiesCountCallback} from "../../Typedef/Callbacks";
+import {IRavenObject} from "../../Typedef/IRavenObject";
+import {DocumentConstructor, DocumentConventions, DocumentType} from "../Conventions/DocumentConventions";
+import {ConditionValue, OrderingType, QueryOperator, SearchOperator} from "./Query/QueryLanguage";
+import {SpatialCriteria} from "./Query/Spatial/SpatialCriteria";
+import {WhereParams, IWhereParams} from "./Query/WhereParams";
+import {SpatialUnit} from "./Query/Spatial/SpatialUnit";
+import {SpatialRelation} from "./Query/Spatial/SpatialRelation";
+import {IOptionsSet} from "../../Typedef/IOptionsSet";
+import {IndexQuery} from "../../Database/Indexes/IndexQuery";
 
 export interface IDocumentQueryOptions<T> {
-  documentType?: DocumentType<T>, 
+  documentType?: DocumentType<T>,
   indexName?: string;
-  usingDefaultOperator?: QueryOperator;
-  waitForNonStaleResults?: boolean;
-  includes?: string[];
   nestedObjectTypes?: IRavenObject<DocumentConstructor>;
   withStatistics?: boolean;
+  indexQueryOptions?: IOptionsSet;
 }
 
-export interface IDocumentQuery<T> {
-  or: IDocumentQuery<T>;
-  and: IDocumentQuery<T>;
+export interface IDocumentQuery<T extends Object = IRavenObject> {
   not: IDocumentQuery<T>;
-  selectFields(...args: string[]): IDocumentQuery<T>;
-  search(fieldName: string, searchTerms: string | string[], escapeQueryOptions?: EscapeQueryOption, boost?: number): IDocumentQuery<T>;
-  where(conditions: IDocumentQueryConditions): IDocumentQuery<T>;
-  whereEquals<V extends LuceneValue>(fieldName: string, value: V, escapeQueryOptions?: EscapeQueryOption): IDocumentQuery<T>;
-  whereEndsWith(fieldName: string, value: string): IDocumentQuery<T>;
-  whereStartsWith(fieldName: string, value: string): IDocumentQuery<T>;
-  whereIn<V extends LuceneValue>(fieldName: string, values: V[]): IDocumentQuery<T>;
-  whereBetween<V extends LuceneValue>(fieldName: string, start?: V, end?: V): IDocumentQuery<T>;
-  whereBetweenOrEqual<V extends LuceneValue>(fieldName: string, start?: V, end?: V): IDocumentQuery<T>;
-  whereGreaterThan<V extends LuceneValue>(fieldName: string, value: V): IDocumentQuery<T>;
-  whereGreaterThanOrEqual<V extends LuceneValue>(fieldName: string, value: V): IDocumentQuery<T>;
-  whereLessThan<V extends LuceneValue>(fieldName: string, value: V): IDocumentQuery<T>;
-  whereLessThanOrEqual<V extends LuceneValue>(fieldName: string, value: V): IDocumentQuery<T>;
-  whereIsNull(fieldName: string): IDocumentQuery<T>;
-  whereNotNull(fieldName: string): IDocumentQuery<T>;
-  orderBy(fieldsNames: string | string[]): IDocumentQuery<T>;
-  orderByDescending(fieldsNames: string | string[]): IDocumentQuery<T>;
+  indexName: string;
+  collectionName: string;
+  conventions: DocumentConventions;
+  isFetchingAllFields: boolean;
+  isDynamicMapReduce: boolean;
+  rawQuery(query: string): IDocumentQuery<T>;
```

Comment:
This method is not the part of this interface in C# client anymore. Please move that to `session.advanced.rawQuery()` :
https://github.com/ravendb/ravendb/blob/4ef25ef622da636cdda4f375b6d1c86b38ca408d/src/Raven.Client/Documents/Session/DocumentSession.RawQuery.cs

Reply:
we do not have yet Session.Advanced in nodejs client

Reply:
also the code you provided is just an wrapper which calls RawQuery() method, so anyway this method should be in DocumentQuery and should be public

Reply:
ok, can you create it and add only that method there?

Also please note that method (`session.advanced.rawQuery()`) does should not return an interface allowing to modify the query, so please add another query interface for this (https://github.com/ravendb/ravendb/blob/v4.0/src/Raven.Client/Documents/Session/IDocumentQuery.cs#L66) preventing from modifying it. Only stuff like (taken from C# impl):
```
AddParameter
Conventions
AfterQueryExecuted
AfterStreamExecuted
BeforeQueryExecuted
InvokeAfterQueryExecuted
InvokeAfterStreamExecuted
NoCaching
NoTracking
ShowTimings
Skip
Statistics
Take
UsingDefaultOperator
WaitForNonStaleResults
WaitForNonStaleResultsAsOf
GetIndexQuery
Conventions
First
FirstOrDefault
Single
SingleOrDefault
Count
```

So in Typescript it's probably just:
```
  indexName: string;
  collectionName: string;
  conventions: DocumentConventions;
  first(callback?: EntityCallback<T>): Promise<T>;
  single(callback?: EntityCallback<T>): Promise<T>;
  all(callback?: QueryResultsCallback<T[]>): Promise<T[]>;
  all(callback?: QueryResultsCallback<QueryResultsWithStatistics<T>>): Promise<QueryResultsWithStatistics<T>>;
  count(callback?: EntitiesCountCallback): Promise<number>;
```

Reply:
done